utils/ms_bot_framework_utils/conversation.py

```diff
@@ -0,0 +1,118 @@
+        self.bot = bot
+        self.bot_id = activity['recipient']['id']
+        self.bot_name = activity['recipient']['name']
+        self.service_url = activity['serviceUrl']
+        self.channel_id = activity['channelId']
+        self.conversation_id = activity['conversation']['id']
+        self.buffer = []
+        self.expect = []
+        self.multiargument_initiated = False
+
+        if self.channel_id not in self.bot.http_sessions.keys() or not self.bot.http_sessions['self.channel_id']:
+            self.bot.http_sessions['self.channel_id'] = requests.Session()
+
+        self.http_session = self.bot.http_sessions['self.channel_id']
+
+        self.handled_activities = {
+            'message': self._handle_message
+        }
+
+    def handle_activity(self, activity: dict):
+        activity_type = activity['type']
+        activity_id = activity['id']
+        log.debug(f'Received activity. Type: {activity_type}, id: {activity_id}')
+
+        if activity_type in self.handled_activities.keys():
+            self.handled_activities[activity_type](activity)
+
+    def _send_activity(self, url: str, out_activity: dict):
+        authorization = f"{self.bot.access_info['token_type']} {self.bot.access_info['access_token']}"
+        headers = {
+            'Authorization': authorization,
+            'Content-Type': 'application/json'
+        }
+
+        response = self.http_session.post(
+            url=url,
+            json=out_activity,
+            headers=headers)
+
+        log.debug(f'Sent activity to the MSBotFramework server. '
+                  f'Response code: {response.status_code}, response contents: {response.json()}')
+
+    def _send_message(self, message_text: str, in_activity: dict = None):
+        service_url = self.service_url
+
+        out_activity = {
+            'type': 'message',
+            'from': {
+                'id': self.bot_id,
+                'name': self.bot_name
+            },
+            'conversation': {
+                'id': self.conversation_id
+            },
+            'text': message_text
+        }
+
+        if in_activity:
+            try:
+                service_url = in_activity['serviceUrl']
+            except KeyError:
+                pass
+
+            try:
+                out_activity['recepient']['id'] = in_activity['from']['id']
+            except KeyError:
+                pass
+
+            try:
+                out_activity['conversation']['name'] = in_activity['conversation']['name']
+            except KeyError:
+                pass
+
+            try:
+                out_activity['recepient']['name'] = in_activity['from']['name']
+            except KeyError:
+                pass
+
+        url = urljoin(service_url, f"v3/conversations/{self.conversation_id}/activities")
+
+        self._send_activity(url, out_activity)
+
+    def _handle_message(self, in_activity: dict):
+        if 'text' in in_activity.keys():
```

Comment:
And what if `'text'` is not there? Maybe log the message to see what it was? I would also reply something about an unsupported message type so that the user would get some response

Reply:
Done